### ISSUE TITLE
Possible bogus items from Dataset.iter_chunks()

### ISSUE THREAD

**Issue by ivilata** [h5py/h5py#2341]
Hi! With h5py 3.10.0 (built from source), native HDF5 1.14.0 & Python 3.10 on GNU Guix AMD64, some calls to `Dataset.iter_chunks()` yield bogus slices with empty/negative sizes.  Example session:

```
$ python
Python 3.10.7 (main, Jan  1 1970, 00:00:01) [GCC 11.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import h5py, numpy
>>> print(h5py.version.info)
Summary of the h5py configuration
---------------------------------

h5py    3.10.0
HDF5    1.14.0
Python  3.10.7 (main, Jan  1 1970, 00:00:01) [GCC 11.3.0]
sys.platform    linux
sys.maxsize     9223372036854775807
numpy   1.26.1
cython (built with) 0.29.36
numpy (built against) 1.21.6
HDF5 (built against) 1.14.0

>>> a = numpy.arange(25).reshape(5, 5)
>>> h5f = h5py.File('iterchunks.h5', 'w')
>>> d = h5f.create_dataset('test1', data=a, chunks=(2, 2))
>>> d[:]
array([[ 0,  1,  2,  3,  4],
       [ 5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14],
       [15, 16, 17, 18, 19],
       [20, 21, 22, 23, 24]])
>>> d[3:, 3:]
array([[18, 19],
       [23, 24]])
>>> for s in d.iter_chunks((slice(3, 5, None), slice(3, 5, None))):
...     print(s)
...
(slice(3, 4, 1), slice(3, 4, 1))
(slice(3, 4, 1), slice(4, 5, 1))
(slice(4, 5, 1), slice(3, 2, 1))
(slice(4, 5, 1), slice(3, 4, 1))
(slice(4, 5, 1), slice(4, 5, 1))
```

The 3rd line printed by the loop shows a slice with start < stop in the 2nd dimension. In fact, that entry should not have been produced at all, as the rest of items yielded do cover the desired slice (`[[18, 19], [23, 24]]`). Calling `iter_chunks()` with no slice seems to work just fine, however, listing the 9 chunks covering the whole dataset:

```
>>> for s in d.iter_chunks():
...     print(s)
...
(slice(0, 2, 1), slice(0, 2, 1))
(slice(0, 2, 1), slice(2, 4, 1))
(slice(0, 2, 1), slice(4, 5, 1))
(slice(2, 4, 1), slice(0, 2, 1))
(slice(2, 4, 1), slice(2, 4, 1))
(slice(2, 4, 1), slice(4, 5, 1))
(slice(4, 5, 1), slice(0, 2, 1))
(slice(4, 5, 1), slice(2, 4, 1))
(slice(4, 5, 1), slice(4, 5, 1))
```

The same behaviour also happens with Guix' native `python-h5py` (version 3.8.0) package.

Please indicate if I can help debug the issue further in any way.

Thanks! :smiley:

**Comment by takluyver** [comment]
For anyone who'd like to work on this, you can find the code behind `iter_chunks()` here:

https://github.com/h5py/h5py/blob/de331aee7f60feca9035466fa9203027e4f75365/h5py/_hl/dataset.py#L322-L393

**Comment by Delengowski** [comment]
I'm going to take a crack at this one.

**Comment by Delengowski** [comment]
I've spent a little time trying to understanding the problem. I understand the problem, but am failing to see the pattern of the incrementation of `ChunkIterator._chunk_index`, its not making much sense to me, and I need to look at it a bit more.

I would expected the chunk index to increment like

[0,0]
[0,1]
[0,2]
[1,0]
[1,1]
[1,2]
[2,0]
[2,1]
[2,2]

but you run the below it goes up to 3...not that this is wrong

but it seems like the chunk index incrementation is the issue, its not taking a large enough step in the case that OP presents and is what causes the bad slice to be created.

```
chunks = d.iter_chunks()
ci = chunks._chunk_index
for x in chunks:
    print(ci, x, "\n", a[x])
    ci = chunks._chunk_index

chunks = d.iter_chunks((slice(3, 5, None), slice(3, 5, None)))
ci = chunks._chunk_index
for x in chunks:
    print(ci, x, "\n", a[x])
    ci = chunks._chunk_index
```

**Comment by Delengowski** [comment]
I've been looking over this to understand the problem. I believe the problem resides in the `while` loop of `__next__` in `ChunkIterator`. It seems like the way it increments `_chunk_index` is linearly and its just not incrementing enough given the passed selection. The chunk residing at index `[2,0]` should be skipped completely.

![image](https://github.com/h5py/h5py/assets/36213781/93ce11da-a669-47b5-ae38-889d0693685c)

**Comment by takluyver** [comment]
Fixed by #2381
